DigitalLearningSolutions.Data/Services/LearningHubAuthApiUrlService.cs

```diff
@@ -0,0 +1,77 @@
+﻿namespace DigitalLearningSolutions.Data.Services
+{
+    using System.Web;
+    using DigitalLearningSolutions.Data.Helpers;
+    using Microsoft.Extensions.Configuration;
+
+    public interface ILearningHubAuthApiUrlService
+    {
+        string GetLoginUrlForDelegateAuthIdAndResourceUrl(string resourceUrl, int delegateAuthId);
+
+        string GetLinkingUrlForResource(int resourceReferenceId);
+    }
+
+    public class LearningHubAuthApiUrlService : ILearningHubAuthApiUrlService
+    {
+        private readonly IConfiguration config;
+        private readonly IGuidService guidService;
+        private readonly ILearningHubSsoSecurityService learningHubSsoSecurityService;
+
+        public LearningHubAuthApiUrlService(
+            IConfiguration config,
+            IGuidService guidService,
+            ILearningHubSsoSecurityService learningHubSsoSecurityService
+        )
+        {
+            this.config = config;
+            this.guidService = guidService;
+            this.learningHubSsoSecurityService = learningHubSsoSecurityService;
+        }
+
+        private string AuthBaseUrl => config.GetLearningHubAuthApiBaseUrl();
+        private string ClientCode => config.GetLearningHubAuthApiClientCode();
+
+        public string GetLoginUrlForDelegateAuthIdAndResourceUrl(string resourceUrl, int delegateAuthId)
+        {
+            var idHash = learningHubSsoSecurityService.GenerateHash(delegateAuthId.ToString());
+            var loginQueryString =
+                ComposeLoginQueryString(ClientCode, delegateAuthId, idHash, resourceUrl);
+
+            var loginEndpoint = config.GetLearningHubAuthApiLoginEndpoint();
+
+            return AuthBaseUrl + loginEndpoint + loginQueryString;
+        }
+
+        public string GetLinkingUrlForResource(int resourceReferenceId)
+        {
+            var state = ComposeCreateUserState(resourceReferenceId);
+            var stateHash = learningHubSsoSecurityService.GenerateHash(state);
+            var createUserQueryString = ComposeCreateUserQueryString(ClientCode, state, stateHash);
+
+            var linkingEndpoint = config.GetLearningHubAuthApiLinkingEndpoint();
+
+            return AuthBaseUrl + linkingEndpoint + createUserQueryString;
+        }
+
+        private string ComposeCreateUserState(int resourceReferenceId)
+        {
+            return guidService.NewGuid() + $"_refId:{resourceReferenceId}";
+        }
+
+        private static string ComposeLoginQueryString(
+            string clientCode,
+            int? learningHubAuthId,
+            string idHash,
+            string resourceUrl
+        )
+        {
+            var encodedUrl = HttpUtility.UrlEncode(resourceUrl);
+            return $"?clientcode={clientCode}&userid={learningHubAuthId}&hash={idHash}&endclientUrl={encodedUrl}";

```

Comment:
In the documentation these query param keys are `pascal-case`.

Reply:
This doesn't appear to be the case in the document I'm looking at ('Learning Hub External Single Sign On.docx' from HEEDLS-676), is my version out of date or not the authoritative source?

Reply:
There is a version control table on page 2. Mine is at `v1.1`.